File: Java and Kotlin/Other/FFT.java
import java.util.Arrays;

public class FFT {
    
    static class Complex {
        private final double real;
        private final double imaginary;
        
        public Complex(double real, double imaginary) {
            this.real = real;
            this.imaginary = imaginary;
        }
        
        public double getReal() {
            return real;
        }
        
        public double getImaginary() {
            return imaginary;
        }
        
        public Complex add(Complex other) {
            return new Complex(this.real + other.real, this.imaginary + other.imaginary);
        }
        
        public Complex subtract(Complex other) {
            return new Complex(this.real - other.real, this.imaginary - other.imaginary);
        }
        
        public Complex multiply(Complex other) {
            double r = this.real * other.real - this.imaginary * other.imaginary;
            double i = this.real * other.imaginary + this.imaginary * other.real;
            return new Complex(r, i);
        }
        
        public Complex divide(Complex other) {
            double denominator = other.real * other.real + other.imaginary * other.imaginary;
            if (denominator == 0) {
                throw new ArithmeticException("Division by zero complex number");
            }
            
            double r = (this.real * other.real + this.imaginary * other.imaginary) / denominator;
            double i = (this.imaginary * other.real - this.real * other.imaginary) / denominator;
            return new Complex(r, i);
        }
        
        public Complex conjugate() {
            return new Complex(real, -imaginary);
        }
        
        public double magnitude() {
            return Math.sqrt(real * real + imaginary * imaginary);
        }
        
        public static Complex exp(Complex exponent) {
            if (exponent.real != 0) {
                throw new IllegalArgumentException("This exp implementation only works for purely imaginary exponents");
            }
            double theta = exponent.imaginary;
            return new Complex(Math.cos(theta), Math.sin(theta));
        }
        
        @Override
        public String toString() {
            if (imaginary >= 0) {
                return String.format("%.5f+%.5fi", real, imaginary);
            } else {
                return String.format("%.5f%.5fi", real, imaginary);
            }
        }
    }
    
    public static Complex[] computeFFT(Complex[] x) {
        int N = x.length;
        
        if ((N & (N - 1)) != 0) {
            int nextPowerOf2 = 1;
            while (nextPowerOf2 < N)
                nextPowerOf2 <<= 1;
            x = Arrays.copyOf(x, nextPowerOf2);
            N = x.length;
        }
        
        if (N <= 1)
            return x;
        
        Complex[] even = new Complex[N / 2];
        Complex[] odd = new Complex[N / 2];
        
        for (int i = 0; i < N / 2; i++) {
            even[i] = x[2 * i];
            odd[i] = x[2 * i + 1];
        }
        
        Complex[] fftEven = computeFFT(even);
        Complex[] fftOdd = computeFFT(odd);
        
        Complex[] result = new Complex[N];
        for (int k = 0; k < N / 2; k++) {
            Complex t = Complex.exp(new Complex(0, -2 * Math.PI * k / N)).multiply(fftOdd[k]);
            result[k] = fftEven[k].add(t);
            result[k + N / 2] = fftEven[k].subtract(t);
        }
        
        return result;
    }
    
    public static Complex[] computeInverseFFT(Complex[] X) {
        int N = X.length;
        Complex[] XConj = new Complex[N];
        for (int i = 0; i < N; i++) {
            XConj[i] = X[i].conjugate();
        }
        
        Complex[] result = computeFFT(XConj);
        
        Complex[] normalized = new Complex[N];
        for (int i = 0; i < N; i++) {
            normalized[i] = result[i].conjugate().divide(new Complex(N, 0));
        }
        
        return normalized;
    }
    
    public static double[] polynomialMultiply(double[] poly1, double[] poly2) {
        int n = poly1.length + poly2.length - 1;
        int nextPowerOf2 = 1;
        while (nextPowerOf2 < n)
            nextPowerOf2 <<= 1;
        
        double[] p1 = Arrays.copyOf(poly1, nextPowerOf2);
        double[] p2 = Arrays.copyOf(poly2, nextPowerOf2);
        
        Complex[] cp1 = new Complex[nextPowerOf2];
        Complex[] cp2 = new Complex[nextPowerOf2];
        for (int i = 0; i < nextPowerOf2; i++) {
            cp1[i] = new Complex(p1[i], 0);
            cp2[i] = new Complex(p2[i], 0);
        }
        
        Complex[] fft1 = computeFFT(cp1);
        Complex[] fft2 = computeFFT(cp2);
        
        Complex[] fftProduct = new Complex[nextPowerOf2];
        for (int i = 0; i < nextPowerOf2; i++) {
            fftProduct[i] = fft1[i].multiply(fft2[i]);
        }
        
        Complex[] result = computeInverseFFT(fftProduct);
        
        double[] realResult = new double[n];
        for (int i = 0; i < n; i++) {
            realResult[i] = Math.round(result[i].getReal() * 1e10) / 1e10;
        }
        
        return realResult;
    }
    
    public static void main(String[] args) {
        System.out.println("=== Fast Fourier Transform Demo ===\n");
        
        System.out.println("Example 1: Computing FFT of [1, 2, 3, 4]");
        Complex[] x = { 
            new Complex(1, 0), 
            new Complex(2, 0), 
            new Complex(3, 0), 
            new Complex(4, 0) 
        };
        Complex[] X = computeFFT(x);
        System.out.println("Input: [1, 2, 3, 4]");
        System.out.print("FFT:  [");
        for (int i = 0; i < X.length; i++) {
            if (i > 0) System.out.print(", ");
            System.out.print(X[i]);
        }
        System.out.println("]");
        
        Complex[] xRecovered = computeInverseFFT(X);
        System.out.print("IFFT: [");
        for (int i = 0; i < xRecovered.length; i++) {
            if (i > 0) System.out.print(", ");
            System.out.printf("%.5f", xRecovered[i].getReal());
        }
        System.out.println("]\n");
        
        System.out.println("Example 2: Polynomial multiplication using FFT");
        System.out.println("Multiplying (2x^2 + 3x + 1) and (x^2 + 2x + 4)");
        double[] poly1 = { 2, 3, 1 };
        double[] poly2 = { 1, 2, 4 };
        
        double[] result = polynomialMultiply(poly1, poly2);
        System.out.println("First polynomial:  " + Arrays.toString(poly1));
        System.out.println("Second polynomial: " + Arrays.toString(poly2));
        System.out.println("Product:           " + Arrays.toString(result));
        System.out.println("(Should be: 2x^4 + 7x^3 + 15x^2 + 14x + 4 -> [2, 7, 15, 14, 4])\n");
        
        System.out.println("Example 3: Frequency analysis of a signal");
        int samples = 8;
        Complex[] signal = new Complex[samples];
        for (int i = 0; i < samples; i++) {
            double t = (double)i / samples;
            double value = Math.sin(2 * Math.PI * 1 * t) + 0.5 * Math.sin(2 * Math.PI * 3 * t);
            signal[i] = new Complex(value, 0);
            System.out.printf("  t[%d] = %.3f\n", i, value);
        }
        
        Complex[] spectrum = computeFFT(signal);
        System.out.println("\nFrequency spectrum (magnitudes):");
        for (int i = 0; i < spectrum.length; i++) {
            double magnitude = spectrum[i].magnitude();
            System.out.printf("  F[%d] = %.3f\n", i, magnitude);
        }
    }
}

File: Java and Kotlin/Other/FFT.kt


import kotlin.math.*

data class Complex(val real: Double, val imaginary: Double) {
    operator fun plus(other: Complex): Complex {
        return Complex(this.real + other.real, this.imaginary + other.imaginary)
    }
    
    operator fun minus(other: Complex): Complex {
        return Complex(this.real - other.real, this.imaginary - other.imaginary)
    }
    
    operator fun times(other: Complex): Complex {
        val r = this.real * other.real - this.imaginary * other.imaginary
        val i = this.real * other.imaginary + this.imaginary * other.real
        return Complex(r, i)
    }
    
    operator fun div(scalar: Double): Complex {
        return Complex(this.real / scalar, this.imaginary / scalar)
    }
    
    fun conjugate(): Complex {
        return Complex(real, -imaginary)
    }
    
    fun magnitude(): Double {
        return sqrt(real * real + imaginary * imaginary)
    }
    
    override fun toString(): String {
        return if (imaginary >= 0) {
            "%.5f+%.5fi".format(real, imaginary)
        } else {
            "%.5f%.5fi".format(real, imaginary)
        }
    }
    
    companion object {
        fun exp(exponent: Complex): Complex {
            if (exponent.real != 0.0) {
                throw IllegalArgumentException("This exp implementation only works for purely imaginary exponents")
            }
            val theta = exponent.imaginary
            return Complex(cos(theta), sin(theta))
        }
        
        val ZERO = Complex(0.0, 0.0)
        val ONE = Complex(1.0, 0.0)
    }
}

class FFT {
    companion object {
        fun computeFFT(x: List<Complex>): List<Complex> {
            var x = x.toMutableList()
            var N = x.size
            
            if ((N and (N - 1)) != 0) {
                var nextPowerOf2 = 1
                while (nextPowerOf2 < N)
                    nextPowerOf2 = nextPowerOf2 shl 1
                while (x.size < nextPowerOf2) {
                    x.add(Complex.ZERO)
                }
                N = x.size
            }
            
            if (N <= 1)
                return x
            
            val even = mutableListOf<Complex>()
            val odd = mutableListOf<Complex>()
            
            for (i in 0 until N / 2) {
                even.add(x[2 * i])
                odd.add(x[2 * i + 1])
            }
            
            val fftEven = computeFFT(even)
            val fftOdd = computeFFT(odd)
            
            val result = MutableList(N) { Complex.ZERO }
            for (k in 0 until N / 2) {
                val t = Complex.exp(Complex(0.0, -2 * PI * k / N)) * fftOdd[k]
                result[k] = fftEven[k] + t
                result[k + N / 2] = fftEven[k] - t
            }
            
            return result
        }
        
        fun computeInverseFFT(X: List<Complex>): List<Complex> {
            val N = X.size
            val XConj = X.map { it.conjugate() }
            
            val result = computeFFT(XConj)
            
            return result.map { it.conjugate() / N.toDouble() }
        }
        
        fun polynomialMultiply(poly1: List<Double>, poly2: List<Double>): List<Double> {
            var n = poly1.size + poly2.size - 1
            var nextPowerOf2 = 1
            while (nextPowerOf2 < n)
                nextPowerOf2 = nextPowerOf2 shl 1
            
            val p1 = poly1.toMutableList().apply { while (size < nextPowerOf2) add(0.0) }
            val p2 = poly2.toMutableList().apply { while (size < nextPowerOf2) add(0.0) }
            
            val cp1 = p1.map { Complex(it, 0.0) }
            val cp2 = p2.map { Complex(it, 0.0) }
            
            val fft1 = computeFFT(cp1)
            val fft2 = computeFFT(cp2)
            
            val fftProduct = fft1.zip(fft2) { a, b -> a * b }
            
            val result = computeInverseFFT(fftProduct)
            
            return result.take(n).map { round(it.real * 1e10) / 1e10 }
        }
        

        @JvmStatic
        fun main(args: Array<String>) {
            println("=== Fast Fourier Transform Demo ===\n")
            
            println("Example 1: Computing FFT of [1, 2, 3, 4]")
            val x = listOf(
                Complex(1.0, 0.0),
                Complex(2.0, 0.0),
                Complex(3.0, 0.0),
                Complex(4.0, 0.0)
            )
            val X = computeFFT(x)
            println("Input: [1, 2, 3, 4]")
            print("FFT:  [")
            X.forEachIndexed { i, value ->
                if (i > 0) print(", ")
                print(value)
            }
            println("]")
            
            val xRecovered = computeInverseFFT(X)
            print("IFFT: [")
            xRecovered.forEachIndexed { i, value ->
                if (i > 0) print(", ")
                print("%.5f".format(value.real))
            }
            println("]\n")
            
            println("Example 2: Polynomial multiplication using FFT")
            println("Multiplying (2x^2 + 3x + 1) and (x^2 + 2x + 4)")
            val poly1 = listOf(2.0, 3.0, 1.0)
            val poly2 = listOf(1.0, 2.0, 4.0)
            
            val result = polynomialMultiply(poly1, poly2)
            println("First polynomial:  $poly1")
            println("Second polynomial: $poly2")
            println("Product:           $result")
            println("(Should be: 2x^4 + 7x^3 + 15x^2 + 14x + 4 -> [2.0, 7.0, 15.0, 14.0, 4.0])\n")
            
            println("Example 3: Frequency analysis of a signal")
            val samples = 8
            val signal = MutableList(samples) { i ->
                val t = i.toDouble() / samples
                val value = sin(2 * PI * 1 * t) + 0.5 * sin(2 * PI * 3 * t)
                Complex(value, 0.0).also {
                    println("  t[$i] = %.3f".format(value))
                }
            }
            
            val spectrum = computeFFT(signal)
            println("\nFrequency spectrum (magnitudes):")
            spectrum.forEachIndexed { i, freq ->
                val magnitude = freq.magnitude()
                println("  F[$i] = %.3f".format(magnitude))
            }
        }
    }
}